### PR TITLE
[CDF-8458] Entity matching update models, retrieve (byids)

### PR DESCRIPTION
**Pull request by rebwib**
This solves https://cognitedata.atlassian.net/browse/CDF-8458?atlOrigin=eyJpIjoiYzFmNjA0ZjYyYmU2NDAxYWFjMjM4NGZlMmRkODdmZjAiLCJwIjoiaiJ9

@sanderland I was not sure what to return. Hope you can give me feedback on whether we should return a list or a model object.

### CODE REVIEW

File: cognite/experimental/_context_client.py [hunk 1]
```diff
@@ -53,7 +53,7 @@ def _fit_model(self, model_path="/fit", headers=None, **kwargs) -> Union[EntityM
         response = self._camel_post(model_path, json=kwargs, headers=headers)
         return self._MODEL_CLASS._load(response.json(), cognite_client=self._cognite_client)
 
-    def retrieve(self, model_id: int) -> ContextualizationModel:
+    def retrieve(self, model_id: int) -> Union[EntityMatchingModel]:
```
Comment (sanderland) [inline cognite/experimental/_context_client.py:56]: union with what?

**Review by sanderland — CHANGES_REQUESTED** [review]
the update call itself should return a list of updated objects, so you should not need to have to call retrieve either before or after. this should only be a single http call.

the sdk takes List[Union[Asset, AssetUpdate]] which is quite involved with autogeneration magic etc.
I would suggest taking a list of models and just throw the entire model dumps at the update endpoint, and have the update endpoint complain if some fields are set that both change and aren't updateable.   is there something in the api design for this?

File: cognite/experimental/_context_client.py (hunk 1 above)
Comment (rebwib) [reply cognite/experimental/_context_client.py:56]: After we removed resource typing it is alway  `EntityMatchingModel` so I can remove the union.

**Comment by rebwib** [comment]
> the update call itself should return a list of updated objects, so you should not need to have to call retrieve either before or after. this should only be a single http call.
> 
> the sdk takes List[Union[Asset, AssetUpdate]] which is quite involved with autogeneration magic etc.
> I would suggest taking a list of models and just throw the entire model dumps at the update endpoint, and have the update endpoint complain if some fields are set that both change and aren't updateable. is there something in the api design for this?

I tried to follow how the SDK updates for instance an asset when i created this method. However, was a bit unsure about the response. Changed it in my last commit. As of now, it is only possible to update one model with each update-call. The reason was to resemble update in the SDK.

There is a section on update in the API design, https://github.com/cognitedata/service-contracts/blob/master/design_guide/API_design_guide.md#update I tried to follow that when I created the update endpoint.

File: cognite/experimental/_context_client.py [hunk 2]
```diff
@@ -89,6 +89,27 @@ def list_jobs(self) -> ContextualizationModelList:
...
+            key
+            for key in model_attributes.keys()
+            if model_attributes[key] != model_update_attributes[key]
+            and key not in ContextualizationModel().__dict__.keys()
+        ]
+        update_dict = {}
+        for attribute in attributes_update:
+            update_dict[attribute] = {"set": model_update_attributes[attribute]}
+
+        return self._camel_post("/update", json={"items": [{"modelId": model.model_id, "update": update_dict}]}).json()
```
Comment (sanderland) [inline cognite/experimental/_context_client.py:111]: definitely convert this to an EntityMatchingModel again though :)

**Comment by sanderland** [comment]
> > the update call itself should return a list of updated objects, so you should not need to have to call retrieve either before or after. this should only be a single http call.
> > the sdk takes List[Union[Asset, AssetUpdate]] which is quite involved with autogeneration magic etc.
> > I would suggest taking a list of models and just throw the entire model dumps at the update endpoint, and have the update endpoint complain if some fields are set that both change and aren't updateable. is there something in the api design for this?
> 
> I tried to follow how the SDK updates for instance an asset when i created this method. However, was a bit unsure about the response. Changed it in my last commit. As of now, it is only possible to update one model with each update-call. The reason was to resemble update in the SDK.
> 
> There is a section on update in the API design, https://github.com/cognitedata/service-contracts/blob/master/design_guide/API_design_guide.md#update I tried to follow that when I created the update endpoint.

Don't really like the retrieve call, but alright then, efficiency shouldn't be that important here.

File: cognite/experimental/_context_client.py (hunk 2 above)
Comment (rebwib) [reply cognite/experimental/_context_client.py]: See 5d8f473. If that was not correct, it would be great if you could push a fix to this branch.

**Comment by rebwib** [comment]
> > > the update call itself should return a list of updated objects, so you should not need to have to call retrieve either before or after. this should only be a single http call.
> > > the sdk takes List[Union[Asset, AssetUpdate]] which is quite involved with autogeneration magic etc.
> > > I would suggest taking a list of models and just throw the entire model dumps at the update endpoint, and have the update endpoint complain if some fields are set that both change and aren't updateable. is there something in the api design for this?
> > 
> > 
> > I tried to follow how the SDK updates for instance an asset when i created this method. However, was a bit unsure about the response. Changed it in my last commit. As of now, it is only possible to update one model with each update-call. The reason was to resemble update in the SDK.
> > There is a section on update in the API design, https://github.com/cognitedata/service-contracts/blob/master/design_guide/API_design_guide.md#update I tried to follow that when I created the update endpoint.
> 
> Don't really like the retrieve call, but alright then, efficiency shouldn't be that important here.

I didn't really like it either, but wasn't sure how else to do it.

**Comment by sanderland** [comment]
in the sdk it's
```
    def _get_update_properties(cls):
        return [key for key in cls.__dict__.keys() if (not key.startswith("_")) and (not key == "labels")]
```
i.e. everything is updateable and the backend ignores things like id if they're the same

**Comment by rebwib** [comment]
> ef _get_update_properties(cls):
>         return [key for key in cls.__dict__.keys() if (not key.startswith("_")) and (not key == "labels")]

Then we have to change the api as well, since for now it only lets you update name and description. Should we let the user update e.g. feature_type?

**Review by sanderland — APPROVED** [review]

**Pull request merged**